File: csharp/PE/Problems/Problem211.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PE.Problems
{
    class Problem211:Problem
    {
        public override int Number
        {
            get { return 211; }
        }

        public override ValueType Solve()
        {
            return 1 + Enumerable.Range(2, 63999998).AsParallel().Where(n=>ips(sosod((uint)n))).Sum();
        }

        private bool ips(ulong n)
        {
            return Math.Sqrt(n)%1 == 0;
        }

        private ulong sosod(uint n)
        {
            ulong s = 0;

            for(uint f=1;f*f<=n;f++)
            {
                if(n%f==0)
                {
                    s += f*f;
                    if (f * f != n)
                        s += (n/f)*(n/f);
                }
            }

            return s;
        }
    }
}

File: cpp/Problem211.cpp
#include <iostream>
using namespace std;

unsigned long sosod(int n);
bool iPS(unsigned long);

int main(){
	const int N = 64000000;
	cout << sosod(N) << endl;
	cout << iPS(N) << endl;

	int s = 1;

	for(int i=2;i<N;i++){
		if(iPS(sosod(i))){
			s+=i;
			cout << i << " " << sosod(i) << endl;
		}
	}

	cout << s << endl;

	return 0;
}

unsigned long sosod(int n){
	unsigned long s = 1;
	unsigned int s2 = 4;

	int i;
	for(i=2;s2<=n;i++,s2=i*i){
		s+=s2;
		if(s2<n)
			s+=(n/i)*(n/i);

	}

	return s+n*n;
}

bool iPS(unsigned long n){
	unsigned long f1 = 1;
	unsigned long f2 = n/f1;

	while(f1<f2){
		f2 = (f1+f2)/2;
		f1 = n/f2;
	}

	return f1==f2;
}
